src/adapter/src/catalog.rs

```diff
@@ -3698,39 +3723,113 @@ impl<S: Append> Catalog<S> {
+                .config
+                .logging;
+
+            let log_and_view_ids = logging.source_and_view_ids();
+            let view_ids: HashSet<_> = logging.view_ids().collect();
+
+            for log_id in log_and_view_ids {
+                // We consider the dependencies of both views and logs, but remove the
+                // views itself. The views are included as they depend on the source,
+                // but we dont need an explicit Op::DropItem as they are dropped together
+                // with the replica.
+                ids_to_drop.extend(
+                    self.get_entry(&log_id)
+                        .used_by()
+                        .into_iter()
+                        .filter(|x| !view_ids.contains(x)),
+                )
+            }
+
+            replicas_to_drop.push((instance.id, replica_id));
+        }
+
+        ops.extend(self.drop_items_ops(&ids_to_drop));
+        (replicas_to_drop, ops)
+    }
+
+    pub fn drop_items_ops(&self, ids: &[GlobalId]) -> Vec<Op> {
         let mut ops = vec![];
         let mut seen = HashSet::new();
         for &id in ids {
-            Self::drop_item_cascade(id, &self.state.entry_by_id, &mut ops, &mut seen);
+            self.drop_item_cascade(id, &mut ops, &mut seen);
         }
         ops
     }
 
-    fn drop_schema_items(
-        schema: &Schema,
-        by_id: &BTreeMap<GlobalId, CatalogEntry>,
-        ops: &mut Vec<Op>,
-        seen: &mut HashSet<GlobalId>,
-    ) {
+    fn drop_schema_items(&self, schema: &Schema, ops: &mut Vec<Op>, seen: &mut HashSet<GlobalId>) {
         for &id in schema.items.values() {
-            Self::drop_item_cascade(id, by_id, ops, seen)
+            self.drop_item_cascade(id, ops, seen);
         }
     }
 
-    fn drop_item_cascade(
-        id: GlobalId,
-        by_id: &BTreeMap<GlobalId, CatalogEntry>,
-        ops: &mut Vec<Op>,
-        seen: &mut HashSet<GlobalId>,
-    ) {
+    fn drop_item_cascade(&self, id: GlobalId, ops: &mut Vec<Op>, seen: &mut HashSet<GlobalId>) {
         if !seen.contains(&id) {
             seen.insert(id);
-            for &u in &by_id[&id].used_by {
-                Self::drop_item_cascade(u, by_id, ops, seen)
+            for u in &self.state.entry_by_id[&id].used_by {
+                self.drop_item_cascade(*u, ops, seen);
             }
-            for subsource in by_id[&id].subsources() {
-                Self::drop_item_cascade(subsource, by_id, ops, seen)
+            for subsource in self.state.entry_by_id[&id].subsources() {
+                self.drop_item_cascade(subsource, ops, seen);
+            }
+            if let Some(linked_cluster_id) =
+                self.state.compute_instances_by_linked_object_id.get(&id)
+            {
+                let name = &self.state.compute_instances_by_id[linked_cluster_id].name;
+                let (_ids, cluster_ops) = self.drop_compute_instance_ops(&[name.clone()]);
+                ops.extend(cluster_ops);
```

Comment:
Being able to look up whether an object is linked to a cluster is what necessitated making these methods rather than associated functions.